File: src/models/shipment.py
"""Shipment Model

Defines the Shipment entity for tracking logistics shipments.
"""

from datetime import datetime
from src.configs.database import db
from sqlalchemy.dialects.postgresql import UUID, JSONB
import uuid

class Shipment(db.Model):
    """Shipment model for logistics tracking"""
    
    __tablename__ = 'shipments'
    
    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tracking_number = db.Column(db.String(100), unique=True, nullable=False, index=True)
    
    # Shipment details
    origin = db.Column(db.String(255), nullable=False)
    destination = db.Column(db.String(255), nullable=False)
    status = db.Column(db.String(50), nullable=False, default='pending')  # pending, in_transit, delivered, delayed, cancelled
    
    # Cargo information
    cargo_description = db.Column(db.Text)
    cargo_weight = db.Column(db.Float)  # in kg
    cargo_value = db.Column(db.Float)  # in USD
    cargo_type = db.Column(db.String(100))
    
    # Carrier information
    carrier_name = db.Column(db.String(255))
    carrier_service = db.Column(db.String(100))
    
    # Timeline
    estimated_departure = db.Column(db.DateTime)
    actual_departure = db.Column(db.DateTime)
    estimated_arrival = db.Column(db.DateTime)
    actual_arrival = db.Column(db.DateTime)
    
    # Current location
    current_location = db.Column(db.String(255))
    current_latitude = db.Column(db.Float)
    current_longitude = db.Column(db.Float)
    
    # Customer information (encrypted)
    customer_name = db.Column(db.String(255))
    customer_email = db.Column(db.String(255))
    customer_phone = db.Column(db.String(50))
    
    # Priority and special handling
    priority = db.Column(db.String(20), default='normal')  # low, normal, high, urgent
    special_instructions = db.Column(db.Text)
    requires_signature = db.Column(db.Boolean, default=False)
    
    # Audit fields
    created_by = db.Column(UUID(as_uuid=True), db.ForeignKey('users.id'))
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Metadata and custom fields
    metadata = db.Column(JSONB, default={})
    
    # Relationships
    events = db.relationship('LogisticsEvent', backref='shipment', lazy='dynamic', cascade='all, delete-orphan')
    predictions = db.relationship('DelayPrediction', backref='shipment', lazy='dynamic', cascade='all, delete-orphan')
    alerts = db.relationship('Alert', backref='shipment', lazy='dynamic', cascade='all, delete-orphan')
    
    def to_dict(self):
        """Convert model to dictionary"""
        return {
            'id': str(self.id),
            'tracking_number': self.tracking_number,
            'origin': self.origin,
            'destination': self.destination,
            'status': self.status,
            'cargo_description': self.cargo_description,
            'cargo_weight': self.cargo_weight,
            'cargo_value': self.cargo_value,
            'cargo_type': self.cargo_type,
            'carrier_name': self.carrier_name,
            'carrier_service': self.carrier_service,
            'estimated_departure': self.estimated_departure.isoformat() + 'Z' if self.estimated_departure else None,
            'actual_departure': self.actual_departure.isoformat() + 'Z' if self.actual_departure else None,
            'estimated_arrival': self.estimated_arrival.isoformat() + 'Z' if self.estimated_arrival else None,
            'actual_arrival': self.actual_arrival.isoformat() + 'Z' if self.actual_arrival else None,
            'current_location': self.current_location,
            'current_latitude': self.current_latitude,
            'current_longitude': self.current_longitude,
            'customer_name': self.customer_name,
            'customer_email': self.customer_email,
            'customer_phone': self.customer_phone,
            'priority': self.priority,
            'special_instructions': self.special_instructions,
            'requires_signature': self.requires_signature,
            'created_at': self.created_at.isoformat() + 'Z' if self.created_at else None,
            'updated_at': self.updated_at.isoformat() + 'Z' if self.updated_at else None,
            'metadata': self.metadata
        }
    
    def __repr__(self):
        return f'<Shipment {self.tracking_number}>'

File: src/models/Shipment.js
/**
 * Shipment Model
 * AI-Powered Logistics Monitoring System
 * 
 * Represents shipment entity with comprehensive validation and security
 * Implements data encryption for sensitive fields and audit logging
 */

const mongoose = require('mongoose');
const crypto = require('crypto');
const { encryptField, decryptField } = require('../utils/encryption');

const shipmentSchema = new mongoose.Schema({
  shipmentId: {
    type: String,
    required: true,
    unique: true,
    index: true,
    default: () => crypto.randomUUID()
  },
  trackingNumber: {
    type: String,
    required: [true, 'Tracking number is required'],
    unique: true,
    index: true,
    trim: true,
    uppercase: true,
    validate: {
      validator: function(v) {
        return /^[A-Z0-9]{10,20}$/.test(v);
      },
      message: 'Tracking number must be 10-20 alphanumeric characters'
    }
  },
  status: {
    type: String,
    required: true,
    enum: ['created', 'in_transit', 'delivered', 'delayed', 'cancelled'],
    default: 'created',
    index: true
  },
  originLocation: {
    address: { type: String, required: true },
    coordinates: {
      latitude: { type: Number, min: -90, max: 90 },
      longitude: { type: Number, min: -180, max: 180 }
    },
    locationCode: String,
    facilityType: String
  },
  destinationLocation: {
    address: { type: String, required: true },
    coordinates: {
      latitude: { type: Number, min: -90, max: 90 },
      longitude: { type: Number, min: -180, max: 180 }
    },
    locationCode: String,
    facilityType: String
  },
  currentLocation: {
    address: String,
    coordinates: {
      latitude: Number,
      longitude: Number
    },
    lastUpdated: { type: Date, default: Date.now }
  },
  transportMode: {
    type: String,
    required: true,
    enum: ['air', 'sea', 'road', 'rail'],
    index: true
  },
  carrier: {
    carrierId: { type: String, required: true, index: true },
    carrierName: { type: String, required: true },
    carrierCode: String,
    contactInfo: {
      phone: String,
      email: String
    }
  },
  timeline: {
    estimatedDeparture: Date,
    actualDeparture: Date,
    estimatedArrival: { type: Date, required: true },
    actualArrival: Date
  },
  cargoDetails: {
    weight: { type: Number, min: 0 },
    volume: { type: Number, min: 0 },
    value: { type: Number, min: 0 },
    description: String,
    hazardous: { type: Boolean, default: false },
    encryptedData: String // Encrypted sensitive cargo information
  },
  customerReference: {
    type: String,
    trim: true
  },
  // PII/PHI fields - encrypted at rest
  customerInfo: {
    customerId: String,
    encryptedName: String,
    encryptedContact: String
  },
  metadata: {
    type: Map,
    of: mongoose.Schema.Types.Mixed
  },
  // Audit fields
  createdBy: {
    type: String,
    required: true
  },
  updatedBy: String,
  deletedBy: String,
  deletedAt: Date,
  // Compliance fields
  dataLineage: [{
    source: String,
    timestamp: Date,
    action: String,
    userId: String
  }],
  consentManagement: {
    dataProcessingConsent: { type: Boolean, default: false },
    consentDate: Date,
    consentVersion: String
  },
  // Security fields
  encryptionVersion: { type: String, default: 'v1' },
  lastSecurityAudit: Date
}, {
  timestamps: true,
  collection: 'shipments'
});

// Indexes for performance optimization
shipmentSchema.index({ status: 1, createdAt: -1 });
shipmentSchema.index({ 'carrier.carrierId': 1, status: 1 });
shipmentSchema.index({ 'timeline.estimatedArrival': 1 });
shipmentSchema.index({ createdAt: -1 });
shipmentSchema.index({ updatedAt: -1 });

// Virtual for tracking URL
shipmentSchema.virtual('trackingUrl').get(function() {
  return `${process.env.BASE_URL}/track/${this.trackingNumber}`;
});

// Pre-save middleware for encryption
shipmentSchema.pre('save', async function(next) {
  try {
    // Encrypt sensitive cargo data
    if (this.isModified('cargoDetails') && this.cargoDetails.value) {
      const sensitiveData = JSON.stringify({
        value: this.cargoDetails.value,
        description: this.cargoDetails.description
      });
      this.cargoDetails.encryptedData = await encryptField(sensitiveData);
    }

    // Encrypt customer PII
    if (this.isModified('customerInfo')) {
      if (this.customerInfo.name) {
        this.customerInfo.encryptedName = await encryptField(this.customerInfo.name);
        delete this.customerInfo.name;
      }
      if (this.customerInfo.contact) {
        this.customerInfo.encryptedContact = await encryptField(this.customerInfo.contact);
        delete this.customerInfo.contact;
      }
    }

    // Update data lineage
    this.dataLineage.push({
      source: 'shipment_model',
      timestamp: new Date(),
      action: this.isNew ? 'create' : 'update',
      userId: this.updatedBy || this.createdBy
    });

    next();
  } catch (error) {
    next(error);
  }
});

// Method to decrypt sensitive data
shipmentSchema.methods.decryptSensitiveData = async function() {
  const decryptedData = {};

  if (this.cargoDetails.encryptedData) {
    const decrypted = await decryptField(this.cargoDetails.encryptedData);
    Object.assign(decryptedData, JSON.parse(decrypted));
  }

  if (this.customerInfo.encryptedName) {
    decryptedData.customerName = await decryptField(this.customerInfo.encryptedName);
  }

  if (this.customerInfo.encryptedContact) {
    decryptedData.customerContact = await decryptField(this.customerInfo.encryptedContact);
  }

  return decryptedData;
};

// Method to sanitize output (remove sensitive fields)
shipmentSchema.methods.toSafeObject = function(userRole) {
  const obj = this.toObject();
  
  // Remove encrypted fields from output
  if (obj.cargoDetails) {
    delete obj.cargoDetails.encryptedData;
  }
  
  if (obj.customerInfo) {
    delete obj.customerInfo.encryptedName;
    delete obj.customerInfo.encryptedContact;
  }

  // Role-based field filtering
  if (userRole !== 'admin' && userRole !== 'logistics_manager') {
    delete obj.cargoDetails.value;
    delete obj.customerInfo;
  }

  return obj;
};

// Static method for secure search
shipmentSchema.statics.secureFind = async function(query, userRole, userId) {
  // Apply role-based access control
  if (userRole === 'customer') {
    query['customerInfo.customerId'] = userId;
  }

  return this.find(query);
};

// Soft delete implementation
shipmentSchema.methods.softDelete = function(userId) {
  this.deletedAt = new Date();
  this.deletedBy = userId;
  this.dataLineage.push({
    source: 'shipment_model',
    timestamp: new Date(),
    action: 'soft_delete',
    userId: userId
  });
  return this.save();
};

const Shipment = mongoose.model('Shipment', shipmentSchema);

module.exports = Shipment;